In [1]:
#importing libraries 
import pandas as pd
import numpy as np
import os

In [2]:
# Import data
path = "/Users/Glebazzz/Jupiter/04.25 Instacart Basket Analysis" 
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'ords_prods_merged_flags.pkl'))

In [3]:
df

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [4]:
ords_prods_merge['user_id'].nunique()

206209

In [5]:
df['user_id'].nunique()

206209

In [6]:
#Renaming some colomns 
df.rename(columns = {'Surnam' : 'Second Name'}, inplace = True)
df.rename(columns = {'STATE' : 'State'}, inplace = True)
df.rename(columns = {'n_dependants' : 'houshold'}, inplace = True)
df.rename(columns = {'income' : 'income USD'}, inplace = True)

In [7]:
#Checking for mixed types
for col in df.columns.tolist():
    if df[col].map(type).nunique() > 1:
        print(f"Mixed types found in column: {col}")

Mixed types found in column: First Name


In [8]:
df.head(100)

user_id  First Name Second Name  Gender       State  Age date_joined  \
0     26711     Deborah    Esquivel  Female    Missouri   48    1/1/2017   
1     33890    Patricia        Hart  Female  New Mexico   36    1/1/2017   
2     65803     Kenneth      Farley    Male       Idaho   35    1/1/2017   
3    125935    Michelle       Hicks  Female        Iowa   40    1/1/2017   
4    130797         Ann     Gilmore  Female    Maryland   26    1/1/2017   
..      ...         ...         ...     ...         ...  ...         ...   
95    86389     Stephen     Murillo    Male     Vermont   69    1/1/2017   
96   102956       Kathy    Mitchell  Female    Missouri   78    1/1/2017   
97   180919  Jacqueline       Roach  Female    Nebraska   21    1/1/2017   
98    74155       Karen        Berg  Female     Alabama   33    1/1/2017   
99    96166         NaN     Oconnor    Male    Oklahoma   51    1/1/2017   

    houshold                        fam_status  income USD  
0          3                           married      165665  
1          0                            single       59285  
2          2                           married       99568  
3          0                            single       42049  
4          1                           married       40374  
..       ...                               ...         ...  
95         3                           married       66800  
96         1                           married       96314  
97         3  living with parents and siblings       56831  
98         3                           married       95285  
99         1                           married      155673  

[100 rows x 10 columns]

In [9]:
df.shape

(206209, 10)

In [10]:
#Find out the the missing values 
df_clean = df.isna().sum()

In [11]:
df_clean.shape

(10,)

In [12]:
df_clean.head(10)

user_id            0
First Name     11259
Second Name        0
Gender             0
State              0
Age                0
date_joined        0
houshold           0
fam_status         0
income USD         0
dtype: int64

Most of the data is clean. Only First Name has missing values — exactly 11,259 missing entries. If I drop the 11259 rows that contain missing values nan in the First _name column from df_customers datarame, it will leave only 194950 user_id. which means I will lose all orders information of the 11259 dropped user_id, because these dropped user_id

In [14]:
# Remove duplicates
df = df.drop_duplicates()

In [15]:
# Drop rows with missing user_id (if any)
df = df[df['user_id'].notna()]

In [16]:
# Drop rows with missing values
df_clean = df.dropna()

In [17]:
# Drop unwanted columns from ords_prods_merge
ords_prods_merge = ords_prods_merge.drop(columns=['_merge', 'Unnamed: 0.1'], errors='ignore')

In [18]:
df_clean.dtypes

user_id         int64
First Name     object
Second Name    object
Gender         object
State          object
Age             int64
date_joined    object
houshold        int64
fam_status     object
income USD      int64
dtype: object

In [19]:
ords_prods_merge.dtypes

product_id                  int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
hour                        int64
days_since_prior_order    float64
add_to_cart_order           int64
reordered                   int64
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                   int64
loyalty_flag               object
avg_value                 float64
spenders_flag              object
frequency                 float64
dtype: object

In [34]:
# Drop missing values
ords_prods_merge = ords_prods_merge.dropna()

In [35]:
#Find out the the missing values 
ords_miss = ords_prods_merge.isna().sum()

In [36]:
ords_miss

product_id                0
product_name              0
aisle_id                  0
department_id             0
prices                    0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
hour                      0
days_since_prior_order    0
add_to_cart_order         0
reordered                 0
price_range_loc           0
busiest_day               0
busiest_days              0
busiest_period_of_day     0
max_order                 0
loyalty_flag              0
avg_value                 0
spenders_flag             0
frequency                 0
dtype: int64

In [44]:
ords_prods_merge_clean = ords_prods_merge.dropna()

In [46]:
#Find out the the missing values 
ords_miss = ords_prods_merge_clean.isna().sum()
ords_miss

product_id                0
product_name              0
aisle_id                  0
department_id             0
prices                    0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
hour                      0
days_since_prior_order    0
add_to_cart_order         0
reordered                 0
price_range_loc           0
busiest_day               0
busiest_days              0
busiest_period_of_day     0
max_order                 0
loyalty_flag              0
avg_value                 0
spenders_flag             0
frequency                 0
dtype: int64

#changing data type
ords_prods_merge_clean.loc[:, 'user_id'] = ords_prods_merge_clean['user_id'].astype('int64')

In [48]:
type(df_clean)

pandas.core.frame.DataFrame

In [50]:
ords_prods_merge_clean = ords_prods_merge_clean.drop(columns=['_merge'], errors='ignore')
df_clean = df_clean.drop(columns=['_merge'], errors='ignore')

In [51]:
# Merge the cleaned orders-products dataframe with the user dataframe
df_merged = ords_prods_merge_clean.merge(
    right=df_clean,
    on='user_id',
    indicator=True,
    how='inner'  
)

In [54]:
df_merged.shape

(28668049, 32)

In [56]:
df_merged.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  2.866805e+07  2.866805e+07   2.866805e+07  2.866805e+07  2.866805e+07   
mean   2.559998e+04  7.119680e+01   9.920124e+00  1.191370e+01  1.710794e+06   
std    1.408181e+04  3.821447e+01   6.282188e+00  4.923769e+02  9.872633e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354200e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.561340e+05   
50%    2.529500e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711162e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565400e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  9.999900e+04  3.421082e+06   

            user_id  order_number  orders_day_of_week          hour  \
count  2.866805e+07  2.866805e+07        2.866805e+07  2.866805e+07   
mean   1.029775e+05  1.822773e+01        2.739956e+00  1.341146e+01   
std    5.941699e+04  1.757197e+01        2.087829e+00  4.244997e+00   
min    1.000000e+00  2.000000e+00        0.000000e+00  0.000000e+00   
25%    5.156300e+04  6.000000e+00        1.000000e+00  1.000000e+01   
50%    1.025870e+05  1.200000e+01        3.000000e+00  1.300000e+01   
75%    1.543770e+05  2.500000e+01        5.000000e+00  1.600000e+01   
max    2.062090e+05  9.900000e+01        6.000000e+00  2.300000e+01   

       days_since_prior_order  add_to_cart_order     reordered     max_order  \
count            2.866805e+07       2.866805e+07  2.866805e+07  2.866805e+07   
mean             1.111055e+01       8.358356e+00  6.298668e-01  3.415928e+01   
std              8.780114e+00       7.135146e+00  4.828402e-01  2.517863e+01   
min              0.000000e+00       1.000000e+00  0.000000e+00  3.000000e+00   
25%              5.000000e+00       3.000000e+00  0.000000e+00  1.400000e+01   
50%              8.000000e+00       6.000000e+00  1.000000e+00  2.800000e+01   
75%              1.500000e+01       1.100000e+01  1.000000e+00  4.800000e+01   
max              3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01   

          avg_value     frequency           Age      houshold    income USD  
count  2.866805e+07  2.866805e+07  2.866805e+07  2.866805e+07  2.866805e+07  
mean   1.186745e+01  1.111059e+01  4.947308e+01  1.501763e+00  9.956262e+04  
std    8.097007e+01  5.881622e+00  1.847837e+01  1.118522e+00  4.311802e+04  
min    1.000000e+00  0.000000e+00  1.800000e+01  0.000000e+00  2.590300e+04  
25%    7.395408e+00  6.759259e+00  3.400000e+01  1.000000e+00  6.719100e+04  
50%    7.829594e+00  9.665893e+00  4.900000e+01  2.000000e+00  9.666100e+04  
75%    8.255046e+00  1.444764e+01  6.500000e+01  3.000000e+00  1.280070e+05  
max    2.500542e+04  3.000000e+01  8.100000e+01  3.000000e+00  5.939010e+05

In [58]:
df_merged['_merge'].value_counts()

_merge
both          28668049
left_only            0
right_only           0
Name: count, dtype: int64

In [60]:
# Export the updated DataFrame
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_merged_customers.pkl'))